mapreader/process/post_process.py

```diff
@@ -0,0 +1,182 @@
+        # reshape to min_x, min_y, max_x, max_y
+        context_grid = [(x[0][0], x[1][0], x[0][1], x[1][1]) for x in context_grid]
+
+        context_list = [
+            self[
+                (self["min_x"] == context_loc[0])
+                & (self["min_y"] == context_loc[1])
+                & (self["max_x"] == context_loc[2])
+                & (self["max_y"] == context_loc[3])
+                & (self["parent_id"] == parent_id)
+            ]
+            for context_loc in context_grid
+        ]
+        context_list = [x.image_id.values[0] for x in context_list if len(x)]
+        return context_list
+
+    def update_preds(self, remap: dict, conf: float = 0.7, inplace: bool = False):
+        """Update the predictions of the chosen patches based on their context.
+
+        Parameters
+        ----------
+        remap : dict
+            A dictionary mapping the old labels to the new labels.
+        conf : float, optional
+            Patches with confidence scores below this value will be relabelled, by default 0.7.
+        inplace : bool, optional
+            Whether to relabel inplace or create new dataframe columns, by default False
+        """
+        if self._label_patches is None:
+            raise ValueError("[ERROR] You must run `get_context` first.")
+        if len(self.context) == 0:
+            raise ValueError(
+                "[ERROR] No patches to update. Try changing which labels you are updating."
+            )
+
+        labels = self._label_patches["predicted_label"].unique()
+        if any([label not in remap.keys() for label in labels]):
+            raise ValueError(
+                f"[ERROR] You must specify a remap for each label in {labels}."
+            )
+
+        # add new label to labels_map if not already present (assume label index is next in sequence)
+        for new_label in remap.values():
+            if new_label not in self.labels_map.values():
+                print(
+                    [
+                        f"[INFO] Adding {new_label} to labels_map at index {len(self.labels_map)}."
+                    ]
+                )
+                self.labels_map[len(self.labels_map)] = new_label
+
+        for ix in tqdm(self.context):
+            self._update_preds_id(
+                ix, labels=labels, remap=remap, conf=conf, inplace=inplace
+            )
+
+    def _update_preds_id(
+        self, ix, labels: str | list, remap: dict, conf: float, inplace: bool = False
+    ):
+        """Update the predictions of the patch with the specified index."""
+        context_list = self.context[ix]
+
+        context_df = self[self["image_id"].isin(context_list)]
+        # drop central patch from context
+        context_df.drop(index=ix, inplace=True)
+
+        # reverse the labels_map dict
+        label_index_dict = {v: k for k, v in self.labels_map.items()}
+
+        prefix = "" if inplace else "new_"
+        if (not any(context_df["predicted_label"].isin(labels))) & (
```

Comment:
I *think* that this line is the filtering rule? If so, might it make sense to separate it as it's own function and give it a name. The name could be `filter_rule_1` with a docstring that explains the logic in plain English like "If patch has label X but no surrounding patches have label X, remove label X from patch"

What you could do then is have an arg for `update_preds` called `filter_func`. You could pass in the filter function when it's called and the filter function would be called on this line. This would mean you could substitute in a different filter function.

So all you have to do to try out a different filter rule is write another small function and call update_preds like `patches.update_preds(filter_func=filter_rule_2 ... other args)`

Reply:
I'm going to make a follow up issue for this as an extra feautre vs within this PR.